assets/javascripts/lib/database.js

```diff
@@ -21,12 +21,74 @@ export let useLocalStorage = false;
 /**
  * Force usage of local storage instead of IndexedDb.
  *
+ * Used in tests
+ *
  * @param {Boolean} value Whether to use local storage.
  */
-export function setUseLocalStorage(value) {
+export function _setUseLocalStorage(value) {
   useLocalStorage = value;
 }
 
+/**
+ * IndexedDb API used to store CryptoKey objects securely
+ */
+export let indexedDb = window.indexedDB;
+
+/**
+ * Sets IndexedDb backend
+ *
+ * Used in tests
+ *
+ * @param {Object} value
+ */
+export function _setIndexedDb(value) {
+  indexedDb = value;
+}
+
+/**
+ * Browser's user agent string
+ */
+export let userAgent = window.navigator.userAgent;
+
+/**
+ * Sets browser's user agent string
+ *
+ * Used in tests
+ *
+ * @param {String} value
+ */
+export function _setUserAgent(value) {
+  userAgent = value;
+}
+
+/**
+ * Warm up IndexedDB to ensure it works normally
+ *
+ * Used in Safari to workaround a bug. indexedDB.open hangs in Safari
+ * if used immediately after page was loaded:
+ * https://bugs.webkit.org/show_bug.cgi?id=226547
+ *
+ * @return {Promise<void>}
+ */
+function initIndexedDb() {
+  const isSafari =
+    !!userAgent.match(/Version\/(\d+).+?Safari/) ||
```

Comment:
I think we should limit this to Safari 14, given that the bug is fixed on 15.

Reply:
You are right. It should be fixed in 15, so I will modify the regex to match only Safari 14.